pac4j-saml: Typecast error in SAML2DefaultResponseValidator

There is a typecast error in SAML2DefaultResponseValidator.validateAssertionSignature().
Look at the following lines:
```java
     final QName role = selfContext.getRole();
        if (signature != null) {
            final String entityId = peerContext.getEntityId();
            validateSignature(signature, entityId, engine);
        } else if (((SPSSODescriptor) role).getWantAssertionsSigned()
                && !peerContext.isAuthenticated()) {
            throw new SAMLException("Assertion or response must be signed");
        }
```
The cast of role to SPSSODescriptor is not valid. Role is of type QName and QName does not have any common part of hierarchy with SPSSODescriptor.

Once the execution gets to the else-if condition (if the signature is null), this code will crash. This stacktrace is from 1.7.1 but valid for 1.8.x too:
```
Caused by: java.lang.ClassCastException: javax.xml.namespace.QName cannot be cast to org.opensaml.saml.saml2.metadata.SPSSODescriptor
	at org.pac4j.saml.sso.impl.SAML2DefaultResponseValidator.validateAssertionSignature(SAML2DefaultResponseValidator.java:659)
	at org.pac4j.saml.sso.impl.SAML2DefaultResponseValidator.validateAssertion(SAML2DefaultResponseValidator.java:400)
	at org.pac4j.saml.sso.impl.SAML2DefaultResponseValidator.validateSamlSSOResponse(SAML2DefaultResponseValidator.java:307)
	at org.pac4j.saml.sso.impl.SAML2DefaultResponseValidator.validate(SAML2DefaultResponseValidator.java:151)
	at org.pac4j.saml.sso.impl.SAML2WebSSOMessageReceiver.receiveMessage(SAML2WebSSOMessageReceiver.java:83)
	at org.pac4j.saml.sso.impl.SAML2WebSSOProfileHandler.receive(SAML2WebSSOProfileHandler.java:54)
	at org.pac4j.saml.client.SAML2Client.retrieveCredentials(SAML2Client.java:253)
	at org.pac4j.saml.client.SAML2Client.retrieveCredentials(SAML2Client.java:77)
	at org.pac4j.core.client.BaseClient.getCredentials(BaseClient.java:222)
	at org.jasig.cas.support.pac4j.web.flow.ClientAction.doExecute(ClientAction.java:196)
	at org.springframework.webflow.action.AbstractAction.execute(AbstractAction.java:188) 
	at org.springframework.webflow.execution.ActionExecutor.execute(ActionExecutor.java:51) 
	... 50 common frames omitted 
```

I only have limited knowledge of OpenSaml but I think we should first find a RoleDescriptor for the QName and then call getWantAssertionsSigned() on it.

Hi,

Thanks for reporting this issue on the SAML support. Would you mind opening
a new Github issue for that?

I planned to release the version 1.8.0 this Thursday, I hope we'll be able
to fix that quickly.

Thanks.
Best regards,
Jérôme


2015-10-06 17:59 GMT+02:00 Jaroslav KAČER <notifications@github.com>:

> There is a typecast error in
> SAML2DefaultResponseValidator.validateAssertionSignature().
> Look at the following lines:
>
>      final QName role = selfContext.getRole();
>         if (signature != null) {
>             final String entityId = peerContext.getEntityId();
>             validateSignature(signature, entityId, engine);
>         } else if (((SPSSODescriptor) role).getWantAssertionsSigned()
>                 && !peerContext.isAuthenticated()) {
>             throw new SAMLException("Assertion or response must be signed");
>         }
>
> The cast of role to SPSSODescriptor is not valid. Role is of type QName
> and QName does not have any common part of hierarchy with SPSSODescriptor.
>
> Once the execution gets to the else-if condition (if the signature is
> null), this code will crash. This stacktrace is from 1.7.1 but valid for
> 1.8.x too:
>
> Caused by: java.lang.ClassCastException: javax.xml.namespace.QName cannot be cast to org.opensaml.saml.saml2.metadata.SPSSODescriptor
>     at org.pac4j.saml.sso.impl.SAML2DefaultResponseValidator.validateAssertionSignature(SAML2DefaultResponseValidator.java:659)
>     at org.pac4j.saml.sso.impl.SAML2DefaultResponseValidator.validateAssertion(SAML2DefaultResponseValidator.java:400)
>     at org.pac4j.saml.sso.impl.SAML2DefaultResponseValidator.validateSamlSSOResponse(SAML2DefaultResponseValidator.java:307)
>     at org.pac4j.saml.sso.impl.SAML2DefaultResponseValidator.validate(SAML2DefaultResponseValidator.java:151)
>     at org.pac4j.saml.sso.impl.SAML2WebSSOMessageReceiver.receiveMessage(SAML2WebSSOMessageReceiver.java:83)
>     at org.pac4j.saml.sso.impl.SAML2WebSSOProfileHandler.receive(SAML2WebSSOProfileHandler.java:54)
>     at org.pac4j.saml.client.SAML2Client.retrieveCredentials(SAML2Client.java:253)
>     at org.pac4j.saml.client.SAML2Client.retrieveCredentials(SAML2Client.java:77)
>     at org.pac4j.core.client.BaseClient.getCredentials(BaseClient.java:222)
>     at org.jasig.cas.support.pac4j.web.flow.ClientAction.doExecute(ClientAction.java:196)
>     at org.springframework.webflow.action.AbstractAction.execute(AbstractAction.java:188)
>     at org.springframework.webflow.execution.ActionExecutor.execute(ActionExecutor.java:51)
>     ... 50 common frames omitted
>
> I only have limited knowledge of OpenSaml but I think we should first find
> a RoleDescriptor for the QName and then call getWantAssertionsSigned() on
> it.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pac4j/pac4j/issues/306>.
>

--- a/pac4j-saml/src/main/java/org/pac4j/saml/sso/impl/SAML2DefaultResponseValidator.java
+++ b/pac4j-saml/src/main/java/org/pac4j/saml/sso/impl/SAML2DefaultResponseValidator.java
@@ -644,14 +644,12 @@ public class SAML2DefaultResponseValidator implements SAML2ResponseValidator {
     protected final void validateAssertionSignature(final Signature signature, final SAML2MessageContext context,
                                                     final SignatureTrustEngine engine) {
 
-        final SAMLSelfEntityContext selfContext = context.getSAMLSelfEntityContext();
         final SAMLPeerEntityContext peerContext = context.getSAMLPeerEntityContext();
 
-        final QName role = selfContext.getRole();
         if (signature != null) {
             final String entityId = peerContext.getEntityId();
             validateSignature(signature, entityId, engine);
-        } else if (((SPSSODescriptor) role).getWantAssertionsSigned()
+        } else if (context.getSPSSODescriptor().getWantAssertionsSignedXSBoolean().getValue()
                 && !peerContext.isAuthenticated()) {
             throw new SAMLException("Assertion or response must be signed");
         }

